### DISCUSSION TITLE
Oobabooga running Llama model with more than 2 threads?

### DISCUSSION THREAD

**Discussion by viperwasp** [oobabooga/text-generation-webui#1589] [General]
I think my 12700K is stuck on 2 threads or something. CPU is at most 22-25% performance. I am running a 30B Alpaca model it is working. But is very slowly.  I am running the alpaca-30b-lora-int4 model named "alpaca-30b-4bit-128g.safetensors"

Here is my command line "python server.py --auto-devices --chat --pre_layer 36 --model alpaca-30b-lora-int4 --wbits 4 --groupsize 128 --model_type llama --threads 16" 

So I am splitting between GPU and CPU. Is it working as intended or is there something I am doing wrong? I want it to run on 16 threads. I know I have 20 threads. Thanks.  I just set threads to 20 and it's 18% CPU use while generating tokens?

**Comment by cparoli** [comment]
Same issue here. Seems like --threads doesnt change anything. With alpaca.cpp I can have good performance with CPU only.

**Reply by viperwasp** [reply]
> Same issue here. Seems like --threads doesnt change anything. With alpaca.cpp I can have good performance with CPU only.

Is this a possible bug or is it intended that CPU only barely uses CPU?

**Comment by edwios** [comment]
Noticed here on a Mac M1 Max, `--threads` --> I am almost certain that it is not being obeyed. However, I am not sure if this is Oobabooga's problem or that of llama-cpp-python's because Oobabooga did initialised Llama model with the correct value to the param `n_threads`, yet the performance when comparing to running `llama.cpp` was so bad that it felt like it was running on a single thread.

**Reply by ossirytk** [reply]
I think llama.cpp thread count is recommended to be the number of logical cores at max. There are some pr's on llama.cpp repo on threading performance improvements, but these seem to be mostly poc and discussion pieces. There does not seem to be a mature enough solution as of yet.
